openedx/Dockerfile

```
@@ -37,52 +53,58 @@ RUN pip uninstall -y ora2 && \
     pip install --src ../venv/src git+https://github.com/regisb/edx-ora2.git@open-release/hawthorn.2#egg=ora2==2.1.17
 
 # Install nodejs requirements
-RUN npm install
+RUN npm set progress=false \
+    && npm install
 ENV PATH ./node_modules/.bin:${PATH}
 
-# Some parts of assets collection can be run without relying on specific
-# settings, nor mounting the staticfiles/ folder
-RUN xmodule_assets common/static/xmodule
-RUN python -c "import pavelib.assets; pavelib.assets.process_npm_assets()"
+# Install private requirements: this is useful for installing custom xblocks.
+# In particular, to install xblocks from a private repository, clone the
+# respositories to ./requirements on the host and add `-e ./myxblock/` to
+# ./requirements/private.txt.
+COPY ./requirements/ /openedx/requirements
+RUN touch /openedx/requirements/private.txt \
+    && pip install --src ../venv/src -r /openedx/requirements/private.txt
```

```suggestion
    && pip install -e -r /openedx/requirements/private.txt
```

If the sources are left in `/openedx/requirements`, the files can be edited "in place" to see changes. Also, duplication can be avoided and some Kbs can be saved by leaving source files where they are.

---

Editing would have to take place inside the container, right? Is that a likely scenario?

As mentioned by email, we want to allow developers to mount edx-platform inside the container, so it makes sense to keep the source outside of the edx-platform folder; pretty much like we would if we had a virtualenv.

---

Sorry, I did not comunicate my vision very well.
Plus, it was based on a misconception I had about docker volumes.

I thought volumes and bind mounts would, on creation, always be pre-filled with the contents of the image.
But TIL this is not the case, and that only happens if you have an unnamed volume.

But still, I stand by my proposed change. I'll try to better state my point.

These are the steps that I think will be needed to work on an xblock:

* When building an image, the desired xblock is cloned into the directory
 `openedx/requiremets`, and referenced from the file `private.txt` in the same
 directory as editable, like this: `-e /openedx/requirements/problem-builder`.

* The image is then built and pushed, and maybe also used in production

* A bug pops up, and someone needs to fix it.

* dev pulls the docker images, _copies_ the contents of `/openedx/requirements` inside
  the image to the local directory `./openedx/requirements` and starts docker compose
  with an option that mounts `./openedx/requirements` to `/openedx/requirements`.

* dev makes changes to the xblock, commits and pushes those changes to some repo

* someone (maybe the dev, maybe a CI job, maybe an operator) builds the docker images
  again and pushes the bug fix to production.

Something similar can happen wrt a theme:

* Pull the relevant docker images that bundle the theme
  (this also gives you an easy way to test changes)

* Copy the themes directory (`/openedx/themes`) from the image to some local dir
  (maybe using the same `./openedx/themes` dir that is used for docker image creation)

* Start the docker compose using a compose file that specifies an additional bind mount
  that will mount `data/themes-src` at `/openedx/themes` inside the container

* Edit the copied theme, commit and push

* Rebuild the docker image with the updated theme, and push the new image version


The main risk that I see with this approach is that some automated script might erase
files the user worked on but did not commit/push.

For this reason I think the automated script should always require the user to delete
everything from a destination directory, before attempting to copy data from a docker
image to some local directory.

What do you think in general about this approach?

---

Let's keep the themes question for later :sweat_smile: Your scenario makes sense for private requirements. Can we agree that:

1. We get rid of `-e` and `--src` for private requirements? We would then have: `pip install -r /openedx/requirements/private.txt` The developers are then free to add `-e` flags inside the requirements.txt file.
2. We keep the `--src` flag for requirements listed in `edx-platform/requirements`

?